### PR TITLE
[mybmw] new binding contribution

### LINKED ISSUE #12117: [myBMW] Add channels for API values Address and InMotion
Binding: MyBMW, created by @weymann (Thanks!)

Please provide channels for the MyBMW API values 'address' (String) and 'InMotion' (Switch)

JSONPATH
$.properties.vehicleLocation.address.formatted
$.properties.inMotion

I realize that inMotion is FALSE all the time for a lot of cars, but there may be models which support this capability.

### PR DESCRIPTION
**Pull request by weymann**
BMW introduced more than 1 year ago their new [MyBMW App](https://www.bmw.de/de/topics/service-zubehoer/bmw-connecteddrive/my-bmw-app.html) for IOS and Android. It's the replacement for the ConnectedDrive App which is covered by the[ BMW ConnectedDrive Binding](https://www.openhab.org/addons/bindings/bmwconnecteddrive/). 

Since some weeks the community [recognizes a decrease of functionality](https://community.openhab.org/t/bmw-connecteddrive-binding/105124/156) in the current provided BMW ConnectedDrive binding. Reason is BMW is shutting down their old API. The new API is based on new servers, different functions and new data structures so the User Interface needs to change. Therefore this new binding will cover the changes.

First drops are [announced and tested in the community](https://community.openhab.org/t/mybmw-binding/130935) and several _data fingerprints_ for different vehicle types are tested in unit tests

Resolves #12117

### CODE REVIEW

File: bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/MyBMWBridgeHandler.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            // check with API call if bridge is online
+            initializerJob = Optional.of(scheduler.schedule(this::requestVehicles, 2, TimeUnit.SECONDS));
+            Bridge b = super.getThing();
+            List<Thing> children = b.getThings();
+            logger.debug("Update {} things", children.size());
+            children.forEach(entry -> {
+                ThingHandler th = entry.getHandler();
+                if (th != null) {
+                    th.dispose();
+                    th.initialize();
```
Comment (fwolter) [inline bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/MyBMWBridgeHandler.java:93]: What's the purpose of this? Actually the framework initializes the child Things when the bridge is initialized.

**Review by fwolter — CHANGES_REQUESTED** [review]
I reviewed 73 / 193 files for now.

File: bundles/org.openhab.binding.mybmw/README.md [hunk 2]
```diff
@@ -0,0 +1,853 @@
...
+```
+
+The debug fingerprint is generated every time the discovery is executed.
+To force a new fingerprint perform the following steps in the UI
+
+1) Go to settings
+2) Select Things entry
+3) Press "+" sign bottom right
+4) Choose the _MyBMW_ binding
+5) Press the _Scan_ button on top
```
[the remark this answers is not in the document]
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/README.md:545]: ok, description is shorter now

File: bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/discovery/VehicleDiscovery.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+     * @param compare String which needs to map with the value
+     * @return String with all field names matching this value separated with Spaces
+     */
+    public static List<String> getObject(Object dto, Object compare) {
+        List<String> l = new ArrayList<String>();
+        for (Field field : dto.getClass().getDeclaredFields()) {
+            try {
+                Object value = field.get(dto);
+                if (compare.equals(value)) {
+                    // buf.append(Converter.capitalizeFirst(field.getName()) + Constants.SPACE);
```
[the remark this answers is not in the document]
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/discovery/VehicleDiscovery.java:215]: removed

**Review by fwolter — CHANGES_REQUESTED** [review]
100 / 194 files viewed

File: bundles/org.openhab.binding.mybmw/src/main/resources/OH-INF/thing/hybrid-range-channel-group.xml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="mybmw"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<channel-group-type id="hybrid-range-values">
+		<label>Mileage, Hybrid Range and Fuel Data</label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.mybmw/src/main/resources/OH-INF/thing/hybrid-range-channel-group.xml:7]: Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions
```suggestion
		<label>Mileage/Hybrid Range/Fuel Data</label>
```

File: bundles/org.openhab.binding.mybmw/src/test/java/org/openhab/binding/mybmw/internal/dto/StatusWrapper.java [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+                        }
+                        assertEquals(wantedMileage, state, "Service Mileage");
+                        break;
+                    default:
+                        assertFalse(true, "Channel " + channelUID + " " + state + " not found");
+                        break;
+                }
+                break;
+            case RANGE_ELECTRIC:
+                assertTrue(isElectric, "Is Eelctric");
```
Comment (fwolter) [inline bundles/org.openhab.binding.mybmw/src/test/java/org/openhab/binding/mybmw/internal/dto/StatusWrapper.java:153]:
```suggestion
                assertTrue(isElectric, "Is Electric");
```

**Review by fwolter — CHANGES_REQUESTED** [review]
I reviewed all files. The code is in a pretty good shape already.

File: bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/RemoteServiceHandler.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    public RemoteServiceHandler(VehicleHandler vehicleHandler, MyBMWProxy myBmwProxy) {
+        handler = vehicleHandler;
+        proxy = myBmwProxy;
+        final VehicleConfiguration config = handler.getConfiguration().get();
+        serviceExecutionAPI = proxy.remoteCommandUrl + config.vin + "/";
+        serviceExecutionStateAPI = proxy.remoteStatusUrl;
+    }
+
+    boolean execute(RemoteService service, String... data) {
+        logger.info("Execute {} with parameters {}", service.getCommand(), data);
```
[the remark this answers is not in the document]
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/RemoteServiceHandler.java:116]: Fixed

File: bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/RemoteServiceHandler.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            }, () -> {
+                logger.warn("No Service executed to get state");
+            });
+            stateJob = Optional.empty();
+        }
+    }
+
+    @Override
+    public void onResponse(@Nullable String result) {
+        logger.info("Remote execution result {}", result);
```
[the remark this answers is not in the document]
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/RemoteServiceHandler.java:164]: Fixed

File: bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/utils/ChargeProfileWrapper.java [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+ * @author Norbert Truchsess - add ChargeProfileActions
+ */
+@NonNullByDefault
+public class ChargeProfileWrapper {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ChargeProfileWrapper.class);
+
+    private static final String CHARGING_WINDOW = "chargingWindow";
+    private static final String WEEKLY_PLANNER = "weeklyPlanner";
+    private static final String ACTIVATE = "activate"; // [todo check variable]
+    private static final String DEACTIVATE = "deactivate"; // [todo check variable]
```
[the remark this answers is not in the document]
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/utils/ChargeProfileWrapper.java:53]: No, usage was checked, simply missed to remove comment

File: bundles/org.openhab.binding.mybmw/src/main/resources/OH-INF/thing/conv-range-channel-group.xml [hunk 9]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="mybmw"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<channel-group-type id="conv-range-values">
+		<label>Combustion Mileage and Range Data</label>
```
[the remark this answers is not in the document]
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/src/main/resources/OH-INF/thing/conv-range-channel-group.xml:7]: done

File: bundles/org.openhab.binding.mybmw/src/main/resources/OH-INF/thing/hybrid-range-channel-group.xml (hunk 4 above)
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/src/main/resources/OH-INF/thing/hybrid-range-channel-group.xml:7]: Group label shortened for all vehicle types

File: bundles/org.openhab.binding.mybmw/src/test/java/org/openhab/binding/mybmw/internal/dto/StatusWrapper.java (hunk 5 above)
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/src/test/java/org/openhab/binding/mybmw/internal/dto/StatusWrapper.java:153]: fixed

File: bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/MyBMWProxy.java [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+    /**
+     * Everything is catched by surroundig try catch
+     * - HTTP Exceptions
+     * - JSONSyntax Exceptions
+     * - potential NullPointer Exceptions
+     *
+     * @return
+     */
+    @SuppressWarnings("null")
+    public synchronized boolean updateToken() {
```
[the remark this answers is not in the document]
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/MyBMWProxy.java:300]: I didn't know this helper class so far. In this particular case it seems BMW enhanced the OAuth2 flow with some additional sugar on top. Using this helper gave me the following results

1) getAuthorizationUrl
delivers the right base address with some parameters in one String
aditional parameters needs to be added manually and auth needs to be done without helper  - which is the major part in this function

2) extractAuthCodeFromAuthResponse
retrun value is delivered with bmw flavored protocol - helper delivers unknown protocol error for
`com.bmw.connected://oauth?client_id=123abc...&response_type=code&...`

3) getAccessTokenResponseByAuthorizationCode
get errors - maybe due to the missing `code_verfier` params used during auth

[the remark this answers is not in the document]
Comment (fwolter) [reply bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/MyBMWProxy.java:300]: I never used the OAuth2 classes in OH. Did you take a look at the other bindings using it?

**Comment by fwolter** [comment]
What is the plan with the BMW ConnectedDrive binding? Is it still usable? Should there be a warning in its readme that it suffers decreased functionality with a reference to this binding?

**Review by fwolter — CHANGES_REQUESTED** [review]
You're relying on the system time zone, but the OH user can configure their time zone within OH. So, instead of relying on the system time zone (`ZonedDateTime.parse(input).withZoneSameInstant(ZoneId.systemDefault())`)  it would be better to use `TimeZoneProvider`. You can enhance this in a follow-up PR, because this is ready to merge if the OAuth2 issue is resolved.

File: bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/MyBMWProxy.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (weymann) [reply bundles/org.openhab.binding.mybmw/src/main/java/org/openhab/binding/mybmw/internal/handler/MyBMWProxy.java:300]: Yes, I scanned binding directory for usage and implemented a test. The result are the above 3 findings from my side which doesn't make it easier in my humble opinion.

**Comment by fwolter** [comment]
Can you fix the conflicting files? Then, this would be ready to merge.

**Comment by weymann** [comment]
For BMW ConnectedDrive I would propose to raise another PR deleting the binding becuase  it isn't working anymore - the login procedure is failing lately so no functionality at all!
The OAuth2 Helper class seems not to help really in the BMW flavored login procedure. As stated above the login URL isn't right, extractingCode doesn't work and also getAccessToken doesn't work from scratch. It seems it doesn't help for code simplicity in this case.

**Review by fwolter — APPROVED** [review]
LGTM

Please don't forget to add a note to the update.lst.

**Pull request merged**